katdal/auth.py

```diff
@@ -0,0 +1,169 @@
+        self._access_token = None
+        self._refresh_token = None
+        self._state = rndstr()
+        self._lock = threading.Lock()
+
+    def _clear_access_token(self):
+        with self._lock:
+            self._access_token = None
+
+    def _ensure_registered(self, provider):
+        if 'redirect_uris' in self._client.registration_response:
+            return
+        self._client.provider_config(provider)
+        client_info = {
+            "client_id": "katdal",
+            "client_secret": "notused",
+            "redirect_uris": ["urn:ietf:wg:oauth:2.0:oob"]
+        }
+        client_reg = RegistrationResponse(**client_info)
+        self._client.store_registration_info(client_reg)
+        self._registered = True
+
+    def _update_access_token(self, response):
+        try:
+            provider = response.headers.get('X-Katdal-Authentication-Provider')
+            if provider is None:
+                provider = urlparse.urljoin(response.request.url, '/')
+            self._ensure_registered(provider)
+            access = None
+
+            if self._refresh_token is not None:
+                logger.debug('Using refresh token to get new access token')
+                access = self._client.do_access_token_refresh(state=self._state, request_args={
+                    "refresh_token": self._refresh_token})
+                if isinstance(access, TokenErrorResponse):
+                    logger.debug('Refresh token failed: %s',
+                                 access.get('error_description', 'unknown error'))
+
+            while not isinstance(access, AccessTokenResponse):
+                nonce = rndstr()
+                auth_req = self._client.construct_AuthorizationRequest(request_args=dict(
+                    client_id=self._client.client_id,
+                    response_type="code",
+                    scope=["openid"],
+                    nonce=nonce,
+                    redirect_uri=self._client.registration_response["redirect_uris"][0],
+                    state=self._state))
+                login_url = auth_req.request(self._client.authorization_endpoint)
+                print('Authentication is required for {}'.format(response.request.url))
+                print('Please visit the URL below:')
+                print()
+                print(login_url)
+                print()
+                code = raw_input('Enter the code: ')
+                auth_json = json.dumps({"state": self._state, "nonce": nonce, "code": code})
+                self._client.parse_response(AuthorizationResponse, info=auth_json,
+                                            sformat="json")
+                access = self._client.do_access_token_request(state=self._state,
+                                                              request_args={"code": code})
+                if isinstance(access, TokenErrorResponse):
+                    print(access.get('error_description', 'unknown error'))
```

Comment:
Does this turn into `StoreUnavailable` eventually, or how are errors reported?

Reply:
It goes around the while loop again and asks the user to try again.